tests/Cluster.py

```diff
@@ -304,25 +349,67 @@ def launch(self, pnodes=1, unstartedNodes=0, totalNodes=1, prodCount=1, topo="me
                 cmdArr.append(path)
 
-        # must be last cmdArr.append before subprocess.call, so that everything is on the command line
-        # before constructing the shape.json file for "bridge"
-        if topo=="bridge":
-            shapeFilePrefix="shape_bridge"
-            shapeFile=shapeFilePrefix+".json"
+        def createDefaultShapeFile(newFile, cmdArr):
             cmdArrForOutput=copy.deepcopy(cmdArr)
             cmdArrForOutput.append("--output")
-            cmdArrForOutput.append(shapeFile)
+            cmdArrForOutput.append(newFile)
             s=" ".join(cmdArrForOutput)
             if Utils.Debug: Utils.Print("cmd: %s" % (s))
             if 0 != subprocess.call(cmdArrForOutput):
-                Utils.Print("ERROR: Launcher failed to create shape file \"%s\"." % (shapeFile))
+                Utils.Print("ERROR: Launcher failed to create shape file \"%s\"." % (newFile))
                 return False
 
-            Utils.Print("opening %s shape file: %s, current dir: %s" % (topo, shapeFile, os.getcwd()))
-            f = open(shapeFile, "r")
-            shapeFileJsonStr = f.read()
+            Utils.Print("opening shape file: %s, current dir: %s" % (newFile, os.getcwd()))
+            f = open(newFile, "r")
+            newFileJsonStr = f.read()
             f.close()
-            shapeFileObject = json.loads(shapeFileJsonStr)
+            return json.loads(newFileJsonStr)
+
+        testnetPrefix = "testnet_"
+        def getTestnetNodeNum(nodeName):
+            p=re.compile(r'^{}(\d+)$'.format(testnetPrefix))
+            m=p.match(nodeName)
+            return int(m.group(1))
+
+        if isinstance(topo, dict):
+            if Utils.Debug: Utils.Print("Creating custom shape topology with the following node connections: {}".format(json.dumps(topo, indent=4, sort_keys=True)))
+            customShapeFile=os.path.join(Utils.ConfigDir, "customShape.json")
+            customShapeFileObject = createDefaultShapeFile(customShapeFile, cmdArr)
+            nodeArray = customShapeFileObject["nodes"]
+            for nodePair in nodeArray:
+                assert(len(nodePair)==2)
+                nodeName=nodePair[0]
+                nodeObject=nodePair[1]
+                if nodeName == "bios":
+                    continue
+                nodeNum=getTestnetNodeNum(nodeName)
+                customShapePeers = []
+                if nodeNum not in topo:
+                    if Utils.Debug: Utils.Print("node name: {} was not included in the topo structure: {}".format(nodeName, topo))
+                    nodeObject["peers"] = customShapePeers
+                    continue
+                peers = topo[nodeNum]
+                Utils.Print("nodeNum: {}, peers: {}".format(nodeNum, peers))
+                for peer in peers:
+                    Utils.Print("peer: {}".format(peer))
+                    assert(peer < totalNodes)
+                    peerStr = str(peer) if peer > 9 else "0" + str(peer)
```

Comment:
Prefer `'{}{:02}'.format(testnetPrefix, peer)` used in the next line directly.

Reply:
ok